File: julia/test/test_parity_fixtures.jl
# V3 cross-language parity (docs/VALIDATION_AND_PERF_PLAN.md): validate the
# Julia kernel against the committed fixtures under <repo_root>/parity.
#
# The fixtures are outputs of the MATLAB V1 oracle (matlab/tests/oracle_bsqr.m,
# a literal transcription of Algorithm 1 with no incremental recurrences),
# written by matlab/tests/generate_parity_fixtures.m with %.17g formatting so
# both languages read bit-identical inputs. Inputs are screened at generation
# time for criterion near-ties, which is what makes exact pivot-sequence
# equality a fair demand across BLAS runtimes. MATLAB consumes the same files
# in matlab/tests/test_parity_fixtures.m.

using DelimitedFiles
using LinearAlgebra: norm

const PARITY_DIR = normpath(joinpath(@__DIR__, "..", "..", "parity"))

read_fixture(name::AbstractString) =
    readdlm(joinpath(PARITY_DIR, name), ',', Float64)

rel_err(X, Y) = norm(X - Y) / max(norm(Y), eps(Float64))

# The fixture zoo sits below the panel kernel's k*n crossover, so the default
# dispatch runs the unblocked kernel. Each fixture is therefore checked twice:
# once at defaults and once with the panel kernel forced, keeping both
# execution paths pinned to the oracle.
const PANEL_MODES = [
    ("default", Dict{String,String}()),
    ("forced panel", Dict("BS_PANEL_NB" => "8", "BS_PANEL_MIN_KN" => "0")),
]

@testset "Cross-language parity fixtures (vs MATLAB oracle, $mode_name)" for (mode_name, mode_env) in PANEL_MODES
    @test isdir(PARITY_DIR)

    manifest = readlines(joinpath(PARITY_DIR, "manifest.csv"))
    @test first(manifest) == "name,m,n,k,rinv_check,gap_min,rtol_R,rtol_rinv,rtol_crit"

    withenv(pairs(mode_env)...) do
    for line in manifest[2:end]
        fields = split(line, ',')
        name = String(fields[1])
        m, n, k = parse.(Int, fields[2:4])
        rinv_check = parse(Int, fields[5]) == 1
        rtol_R = parse(Float64, fields[7])
        rtol_rinv = parse(Float64, fields[8])
        rtol_crit = parse(Float64, fields[9])

        @testset "$name" begin
            A = read_fixture("$(name)_A.csv")
            @test size(A) == (m, n)
            p_exp = Int.(vec(read_fixture("$(name)_p.csv")))
            R_exp = read_fixture("$(name)_R.csv")
            crit_exp = vec(read_fixture("$(name)_crit.csv"))

            F = bsqr(A; k = k, return_rinv_r12 = rinv_check,
                track_inverse_frob = true)

            @test F.ksteps == k
            @test perm(F) == p_exp
            @test rel_err(R(F), R_exp) < rtol_R

            # V4 trace parity: the per-step selected-pivot criterion values
            # (frob_inv_trace) against the oracle's from-scratch trace.
            crit = F.frob_inv_trace
            @test length(crit) == k
            @test maximum(abs.(crit .- crit_exp) ./ crit_exp) < rtol_crit

            if rinv_check
                rinv_exp = read_fixture("$(name)_rinv.csv")
                @test rel_err(rinv_r12(F), rinv_exp) < rtol_rinv
            end
        end
    end
    end
end

File: matlab/tests/test_parity_fixtures.m
function tests = test_parity_fixtures
%TEST_PARITY_FIXTURES V3 validation: committed fixtures vs. MATLAB backends.
%   The fixtures under <repo_root>/parity are oracle outputs written by
%   generate_parity_fixtures. This test asserts that
%     * the committed inputs are bit-identical to the current parity_zoo
%       (catches "zoo changed but fixtures not regenerated" drift);
%     * both MATLAB backends reproduce the expected pivot sequence exactly
%       and the factors to tolerance.
%   The Julia consumer of the same files is julia/test/test_parity_fixtures.jl.
tests = functiontests(localfunctions);
end

function setupOnce(testCase)
repo_root = fileparts(fileparts(fileparts(mfilename('fullpath'))));
addpath(fullfile(repo_root, 'matlab'));
fixdir = fullfile(repo_root, 'parity');
assertTrue(testCase, isfolder(fixdir), ...
    'parity/ fixtures missing; run matlab/tests/generate_parity_fixtures.m');
testCase.TestData.fixdir = fixdir;
testCase.TestData.manifest = readtable(fullfile(fixdir, 'manifest.csv'), ...
    'TextType', 'char');
end

function testFixturesMatchCurrentZoo(testCase)
zoo = parity_zoo();
manifest = testCase.TestData.manifest;
verifyEqual(testCase, height(manifest), numel(zoo), 'zoo size vs manifest');
for idx = 1:numel(zoo)
    z = zoo{idx};
    A_file = read_fixture(testCase, [z.name, '_A.csv']);
    verifyTrue(testCase, isequal(A_file, z.A), ...
        sprintf('%s: committed input differs from parity_zoo; regenerate fixtures', z.name));
end
end

function testMfileMatchesFixtures(testCase)
verify_backend_against_fixtures(testCase, 'mfile');
end

function testMexMatchesFixtures(testCase)
if ~bsqr_mex_available()
    return;
end
verify_backend_against_fixtures(testCase, 'mex');
end

function verify_backend_against_fixtures(testCase, backend)
manifest = testCase.TestData.manifest;
for row = 1:height(manifest)
    name = manifest.name{row};
    k = manifest.k(row);
    n = manifest.n(row);

    A = read_fixture(testCase, [name, '_A.csv']);
    p_exp = read_fixture(testCase, [name, '_p.csv']);
    R_exp = read_fixture(testCase, [name, '_R.csv']);
    crit_exp = read_fixture(testCase, [name, '_crit.csv']);

    [~, R, p, rinv, trace] = bsqr(A, 'backend', backend, 'k', k, ...
        'pivot_format', 'vector', 'return_rinv_r12', true, 'trace', true);

    label = sprintf('%s/%s', name, backend);
    verifyEqual(testCase, p, p_exp, sprintf('%s: pivot sequence', label));
    verifyLessThan(testCase, rel_err(R, R_exp), manifest.rtol_R(row), ...
        sprintf('%s: R factor', label));

    crit_err = max(abs(trace.crit - crit_exp) ./ crit_exp);
    verifyLessThan(testCase, crit_err, manifest.rtol_crit(row), ...
        sprintf('%s: criterion trace', label));

    if manifest.rinv_check(row) && k < n
        rinv_exp = read_fixture(testCase, [name, '_rinv.csv']);
        verifyLessThan(testCase, rel_err(rinv, rinv_exp), manifest.rtol_rinv(row), ...
            sprintf('%s: rinv_r12', label));
    end
end
end

function testTraceParityBetweenBackends(testCase)
% Recompute events are threshold crossings on running norms, so equality
% between the two MATLAB backends pins both the guard logic and the norm
% downdate arithmetic step by step (measured exactly equal on the zoo).
% Pinned to the unblocked kernels: the panel kernel batches guard trips
% per panel (one flag per column), so its counts differ by design — and
% this keeps the unblocked MEX path covered now that panel is the default.
if ~bsqr_mex_available()
    return;
end
old_nb = getenv('BS_PANEL_NB');
restore_nb = onCleanup(@() setenv('BS_PANEL_NB', old_nb));
setenv('BS_PANEL_NB', '0');
manifest = testCase.TestData.manifest;
for row = 1:height(manifest)
    name = manifest.name{row};
    A = read_fixture(testCase, [name, '_A.csv']);
    k = manifest.k(row);
    [~, ~, ~, ~, tm] = bsqr(A, 'backend', 'mfile', 'k', k, ...
        'pivot_format', 'vector', 'trace', true);
    [~, ~, ~, ~, tx] = bsqr(A, 'backend', 'mex', 'k', k, ...
        'pivot_format', 'vector', 'trace', true);
    verifyEqual(testCase, tx.nrecomp, tm.nrecomp, ...
        sprintf('%s: nrecomp trace mfile vs mex', name));
end
end

function X = read_fixture(testCase, fname)
X = readmatrix(fullfile(testCase.TestData.fixdir, fname));
end

function r = rel_err(X, Y)
r = norm(X - Y, 'fro') / max(norm(Y, 'fro'), eps('double'));
end
